_posts/2016-05-13-rustup.md

```diff
@@ -0,0 +1,553 @@
+
+On Linux, download and unpack them with the following commands (the
+output of which is not included here):
+
+```
+rust:~/home$ cd
+rust:~$ curl -O https://dl.google.com/android/android-sdk_r24.4.1-linux.tgz
+rust:~$ tar xzf android-sdk_r24.4.1-linux.tgz
+rust:~$ curl -O http://dl.google.com/android/repository/android-ndk-r10e-linux-x86_64.zip
+rust:~$ unzip android-ndk-r10e-linux-x86_64.zip
+```
+
+We further need to create what the NDK calls a ["standalone
+toolchain"].  We're going to put ours in a directory called
+`android-ndk-r10e`:
+
+["standalone toolchain"]: https://developer.android.com/ndk/guides/standalone_toolchain.html
+
+```
+rust:~$ android-ndk-r10e/build/tools/make-standalone-toolchain.sh \
+      --platform=android-18 --toolchain=arm-linux-androideabi-clang3.6 \
+      --install-dir=android-18-toolchain --ndk-dir=android-ndk-r10e/ --arch=arm
+Auto-config: --toolchain=arm-linux-androideabi-4.8, --llvm-version=3.6
+Copying prebuilt binaries...
+Copying sysroot headers and libraries...
+Copying c++ runtime headers and libraries...
+Copying files to: android-18-toolchain
+Cleaning up...
+Done.
+```
+
+Let's notice a few things about these commands. First, the NDK we
+downloaded, `android-ndk-r10e-linux-x86_64.zip` is not the most recent
+release (which at the time of this writing is 'r11c'). Rust's `std` is
+built against `r10e` and links to symbols that are no longer included
+in the NDK. So for now we have to use the older NDK.  Second, in
+building the standalone toolchain we passed "--platform=android-18" to
+`make-standalone-toolchain.sh`. The "18" here is the Android [API
+level](http://developer.android.com/guide/topics/manifest/uses-sdk-element.html#ApiLevels).
+Today, Rust's `arm-linux-androideabi` target is built against Android
+API level 18, and should theoretically be forwards-compatible with
+subsequent Android API levels. So we're picking level 18 to get the
+greatest Android compatibility that Rust presently allows.
+
+The final thing for us to do is tell Cargo where to find the android
+linker, which is in the standalone NDK toolchain we just created. To
+do that we configure the `arm-linux-androideabi` target in
+[`.cargo/config`] with the 'linker' value. And while we're doing that
+we'll go ahead and set the default target for this project to Android
+so we don't have to keep calling cargo with the `--target` option.
+
+[`.cargo/config`]: http://doc.crates.io/config.html
+
+```toml
+[build]
+target = "arm-linux-androideabi"
+
+[target.arm-linux-androideabi]
+linker = "/home/rust/android-18-toolchain"
+```
+
+Now let's change back to the 'hello' project directory and try
+to build again:
+
+```
+rust:~$ cd hello
+rust:~/hello$ cargo build
+   Compiling hello v0.1.0 (file:///home/rust/hello)
+```
+
+Success! Of course just getting something to build is not the end of
+the story. You've also got to package your code up as an Android
+APK. For that you can use [cargo-apk]. I'm not going to show you that
```

Comment:
I think you can leave off the last sentence here -- just finish with the cargo-apk link.